### ISSUE TITLE
simultaneous controller initiation problem

### ISSUE THREAD

**Issue by shinra358** [Ryochan7/DS4Windows#52]
Xinput allocation problem is back. Two controllers turned on, only one microsoft xbox gamepad is shown in joy.cpl. Unable to turn off 2nd controller's lights. 2nd player unable to play.

**Comment by Ryochan7** [comment]
Looking at the commits for version 1.4.74 shows that nothing regarding Xinput allocation was changed. There is no reason that it would work in 1.4.73 and not the current version. It still works on my system. Something might be going wrong with the Scp Virtual Bus Driver that is outside of DS4Windows' control. If one virtual controller can get allocated then I don't know what can be done; trying to check each virtual port on the hub might not even help.

At some point, I will have to look into the current state of ViGEm and see if it can replace Scp Virtual Bus in future releases. Prior releases of ViGEm seemed to have some problems.

**Comment by shinra358** [comment]
Nah, if i switch back to the previous version, it works fine. Switching
ds4windows should not modify or affect scp drivers. Scp drivers werent
touched at all for that to happen. 2 controllers on, 1 shown in joy.cpl is
what i see. Previous version, 2 controllers on, 2 controllers in joy.cpl.

Compile each update after the previous version as youve included them
chronologically. I will test them all to see where the problem starts.

**Comment by shinra358** [comment]
To slim it down, it's after plugging and then unplugging a controller and then turning that one back on the bluetooth way.

I shoulda said quick charge is messing things up. Because there is no problem with xinput allocation or quick charge in the previous version.

Also, isnt vigem windows 10 only? That would suck for me because I'm on win7 ultimate and don't plan on movie because win10 is far from game friendly as they say.

**Comment by Ryochan7** [comment]
Unfortunately, that situation works fine on my system. I have found two main points where some significant changes were made so I have made builds based on those commits.

Stick Mouse changes (minor code moving regarding lightbar color calculations) (f7d3d1bf479573d08ffe9f4838a4cbe27d52b27a):
https://drive.google.com/open?id=0B6yGHDx0CFzDc1dsV0E1Q203OEE

BT Timeout (f61cd593c5318a253481eb85f0d42c332167902a):
https://drive.google.com/open?id=0B6yGHDx0CFzDTEtMbDJ2UUFYS0U

Just for completeness, here is a current build which also includes some small uncommitted work. 
https://drive.google.com/open?id=0B6yGHDx0CFzDVGhTeGoxVjFvRE0

I think that ViGEm is currently Windows 10 only. As far as I know, the lack of compatibility is only a matter of only having purchased a signing certificate for that platform. Last I remember reading, ViGEm will be released for other versions of Windows in the future. Scp Virtual Bus is not really being maintained anymore and ViGEm does provide some functionality that ScpVBus does not.

**Comment by shinra358** [comment]
This reminds me of the frog on looney tunes that kept pretending to die whenever more than one person was around lol.

I can't reproduce this problem today like yesterday -_-.


But I did find that in .73, .74, and 0603 testbuild that:

1. bluetooth p1 + bluetooth p2 + plug in p2 after = 2 xbox controllers and one generic controller in joy.cpl. So something is up with the hide ds4 feature when this situation happens.

2. when that p2 is unplugged, the controller stays on. It is supposed to turn off when unplugged even if bluetooth was on before. Pretty sure this was fixed before, now it's back.

**Comment by shinra358** [comment]
Everything fixed except quick charge unplug. If battery is full, unplugging turns off the controller which is what we want. If battery is still charging and you unplug, it goes back to bluetooth instead of turning off.

Idle timer light while plugged untested.

**Comment by Ryochan7** [comment]
Maybe the hotplug delay is not long enough. The delay is supposed to be 50 ms and that is supposed to give BT controllers enough time to grab an input report and update the device charging state. With the recent update, I can usually see the BT controller battery status get updated in the GUI before DS4Windows disconnects the controller. Quick Charge works on my system fine under both circumstances. I am thinking about changing how hotplugging is performed in a future update.

The idle light works as intended as well.

**Comment by shinra358** [comment]
The idle light timer dimming until off while plug doesnt wrk for me. Other users are saying that as well in its own thread. I just mentioned it here again bcuz i already mentioned it b4 here.

As for the quick charge while on charging mode, post a test after youve done that tinkering, please.
Thanks for your hard work!

**Comment by Ryochan7** [comment]
The changes will mainly revolve around getting some processing out of the main GUI thread. Testing out the delay hypothesis is still needed. It does seem like there is a small delay between the time a USB cable is plugged into the DS4 and the time that the DS4 sets the value in input reports. Can you try a test build and see if you have better luck with it? The delay has been increased to 200 ms.

https://drive.google.com/open?id=0B6yGHDx0CFzDN1Iwd1NHNWZidlk

**Comment by shinra358** [comment]
That works but seems like this (maybe) makes another problem. When turning on bluetooth controllers one after the other, sometimes, controller colors get shared. Sometimes a controller may stay on a white light instead of the color they are supposed to be. Maybe they are getting confused when theyre trying to connect to the same player number at the same time?

**Comment by Ryochan7** [comment]
The next test is to move quick charge checking to the input thread. That way the delay before running the hotplug routine does not matter anymore; it has been reverted back to 50 ms but it could possibly be removed in future builds. I have tested it and it does work without causing much extra work in the input thread.

https://drive.google.com/open?id=0B6yGHDx0CFzDeHdOUWFudXJveVk

**Comment by shinra358** [comment]
Charging, plugging, then removing turns controller off now. That seems fixed.

But when I turn all 3 controllers on simultaneously, 2 of them go to a solid white light, then eventually turn their proper colors after a short while. But sometimes one of them goes to a solid white light and it stays there. During this time, all 3 xinput controllers are shown in joy.cpl but you can't press home+start to turn the white lighted controller off.

So, quick charge thing is fixed. Them fighting over the positions when turned on simultaneously still remains.

**Comment by Ryochan7** [comment]
New test build. A simpler hotplug technique is being tested.

https://drive.google.com/open?id=0B6yGHDx0CFzDUVRmaTk0bVVzSUE

**Comment by shinra358** [comment]
Previous comment remains the same for both issues.

**Comment by shinra358** [comment]
any update to the simulataneous activation fight over the virtual xinput the controllers have when turned on at the same time? Maybe a controller profile option of "preferred xinput slot for this particular controller' or something like that? So when they fight, each controller will know what place to go to so they don't get confused.

Any other better solution is welcomed as well.

**Comment by Ryochan7** [comment]
Although it might be somewhat unrelated, I found that using the DInput only option in a profile can cause this problem to happen. Multiple calls to the SCP Virtual Bus can potentially happen at the same time and cause conflicts. The issue with DInput only has been fixed in newer commits. I hope that is the last spot where XInput slot issue would occur.

Any type of preferred slot set in the program seems like it would only affect the SCP Virtual Bus. Any application using XInput would still detect controllers in the order they were connected.
